### ISSUE TITLE
[Bug] Chapter 4 – Positional Encoding (RoPE): `rotate_half` uses `split-half` convention but `emb` uses `adjacent-pair` convention

### ISSUE THREAD

**Issue by pavansai018** [raiyanyahya/how-to-train-your-gpt#6]
In the Chapter 4 implementation of Rotary Position Embeddings (_chapters/04_positional_encoding.md_), there is a mismatch between how the frequency embeddings are constructed and how the rotation is applied. This causes incorrect position encoding — the rotation frequencies do not align with the intended dimension pairs.

**What's wrong**

**Step 1** — `emb` is built with `repeat_interleave(2)`

`emb = freqs.repeat_interleave(2, dim=-1)`  # [θ0, θ0, θ1, θ1, θ2, θ2, ...]
This encodes the adjacent-pair convention: dimension pairs are (0,1), (2,3), (4,5), each sharing one frequency.

**Step 2** — `rotate_half ` splits at the midpoint 

```
x1 = x[..., : x.shape[-1] // 2]
x2 = x[..., x.shape[-1] // 2 :]
return torch.cat([-x2, x1], dim=-1)
```
For input **[x0, x1, x2, x3, x4, x5]**, this returns **[-x3, -x4, -x5, x0, x1, x2]**, which implements the split-half convention where pairs are (0,3), (1,4), (2,5).

**Result**: When the rotation formula `x * cos + rotate_half(x) * sin` is applied, dimension 0 (assigned frequency θ0) is paired with dimension 3 (assigned frequency θ1). The frequencies don't match the pairs, so the dot product Q·K no longer encodes purely relative position as the RoPE paper requires.

This is also contradicted by the method's own docstring, which correctly states the output should be [-x1, x0, -x3, x2, -x5, x4] (adjacent-pair), but the implementation does not produce that.

**Proposed fix**

Replace `rotate_half` with an interleaved split to match the adjacent-pair convention used in `emb`:

```
@staticmethod
def rotate_half(x: torch.Tensor) -> torch.Tensor:
    x_even = x[..., 0::2]   # [x0, x2, x4, ...]
    x_odd  = x[..., 1::2]   # [x1, x3, x5, ...]
    return torch.stack((-x_odd, x_even), dim=-1).flatten(-2)
```
produces [-x1, x0, -x3, x2, -x5, x4, ...] — matches docstring
No other changes are needed. `emb = freqs.repeat_interleave(2, dim=-1)` is correct for this convention.

I would like to submit a PR with this fix if the maintainers are open to it.

**Comment by raiyanyahya** [comment]
Thank you for bringing this up. Let me check and verify. I am learning as well :) I have some commits pending and I will add this fix if needed. If you want to contribute please suggest topics that would help the community.

**Comment by raiyanyahya** [comment]
issue fixed with latest commit!
